target-chains/near/receiver/src/governance.rs

```diff
@@ -380,6 +480,223 @@ impl Pyth {
             .then_some(())
-            .ok_or(Error::UnknownSource)
+            .ok_or(UnknownSource)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use {
+        super::*,
+        crate::governance::GovernanceActionId,
+        near_sdk::{
+            test_utils::{
+                accounts,
+                VMContextBuilder,
+            },
+            testing_env,
+        },
+        std::io::{
+            Cursor,
+            Write,
+        },
+    };
+
+    fn get_context() -> VMContextBuilder {
+        let mut context = VMContextBuilder::new();
+        context
+            .current_account_id(accounts(0))
+            .signer_account_id(accounts(0))
+            .predecessor_account_id(accounts(0))
+            .attached_deposit(0)
+            .is_view(false);
+        context
+    }
+
+    #[test]
+    fn test_upgrade() {
+        let mut context = get_context();
+        context.is_view(false);
+        testing_env!(context.build());
+
+        let mut contract = Pyth::new(
+            near_sdk::AccountId::new_unchecked("pyth.near".to_owned()),
+            [0; 32],
+            Source::default(),
+            Source::default(),
+            0.into(),
+            32,
+        );
+
+        contract.codehash = env::sha256(&[1, 2, 3]).try_into().unwrap();
+        contract.upgrade(vec![1, 2, 3]).expect("Upgrade failed");
+    }
+
+    #[test]
+    #[should_panic(expected = "UnauthorizedUpgrade")]
+    fn test_upgrade_fail() {
+        let mut context = get_context();
+        context.is_view(false);
+        testing_env!(context.build());
+
+        let mut contract = Pyth::new(
+            near_sdk::AccountId::new_unchecked("pyth.near".to_owned()),
+            [0; 32],
+            Source::default(),
+            Source::default(),
+            0.into(),
+            32,
+        );
+
+        contract.codehash = env::sha256(&[1, 2, 3]).try_into().unwrap();
+        contract.upgrade(vec![1, 2, 3, 4]).expect("Upgrade failed");
+    }
+
+    #[test]
+    fn test_set_valid_period() {
+        let mut context = get_context();
+        context.is_view(false);
+        testing_env!(context.build());
+
+        let mut contract = Pyth::new(
+            near_sdk::AccountId::new_unchecked("pyth.near".to_owned()),
+            [0; 32],
+            Source::default(),
+            Source::default(),
+            0.into(),
+            32,
+        );
+
+        contract.set_valid_period(100);
+        assert_eq!(contract.stale_threshold, 100);
+    }
+
+    #[test]
+    fn test_set_update_fee() {
+        let mut context = get_context();
+        context.is_view(false);
+        testing_env!(context.build());
+
+        let mut contract = Pyth::new(
+            near_sdk::AccountId::new_unchecked("pyth.near".to_owned()),
+            [0; 32],
+            Source::default(),
+            Source::default(),
+            0.into(),
+            32,
+        );
+
+        contract.set_update_fee(100, 0).expect("Failed to set fee");
```

Comment:
suggest making the exponent nonzero here (it's a little bit better of a test that way)

Reply:
Good spot, fun fact all these tests were auto-generated by Copilot! I thought it was an 8 when I checked them over.